pep/002-mixin-versioning.md

```diff
@@ -0,0 +1,355 @@
+This directory is managed by Porter and may be deleted at any time.
+
+### Porter Lockfile
+
+After Porter resolves the version of a mixin, it records that information in a new Porter lockfile, porter.lock.yaml, that is located next to porter.yaml.
+This file ensures that once a mixin has been resolved, subsequent builds always use the same resolution.
+
+```
+mixins:
+  exec:
+    source: https://cdn.porter.sh/mixins/atom.xml
+    url: https://cdn.porter.sh/mixins/exec/ # URL where the binaries are available
+    version: v0.33.1
+    binaries: # list all available binaries, not just what is used by the current client
+      exec-darwin-amd64:
+        digest: abc123
+      exec-linux-amd64:
+        digest: def456
+```
+
+When deciding if a mixin resolution can be used, Porter looks at the following:
+
+* The mixin name and url must match.
+* The resolved version must match the version range defined in porter.yaml
+
+If the digest of the binaries do not match the lockfile, the build should return an error.
+
+
+### Mixin Commands
+
+We need to repurpose existing mixin commands now that we have a global mixins cache and a local bundle cache.
+Let's review each mixin command and update its signature with a `--global` flag which indicates that the command should operate against the global cache. This flag defaults to false when the command is executed inside a bundle directory.
+
+
+#### porter mixins search
+
+This command does not operate on the cache so no changes are required.
+
+
+#### porter mixins list
+
+When global is false, list the mixins used by the bundle, taking into account the lockfile.
+
+```
+$ porter mixins list [--global=false]
+Name         Version      Source
+exec         v0.33.1      https://cdn.porter.sh/mixins/atom.xml
+```
+
+The json formatted output includes all the information available from the lockfile.
+
+```json
+[
+ {
+    "name": "exec",
+    "source": "https://cdn.porter.sh/mixins/atom.xml",
+    "url": "https://cdn.porter.sh/mixins/exec",
+    "binaries": {
+      "exec-darwin-amd64": {
+        "digest": "abc123",
+      },
+      "exec-linux-amd64": {
+        "digest": "def456",
+      }
+    },
+    "version": "v0.33.1",
+    "commit": "d433a66",
+    "author": "Porter Authors"
+  }
+]
+```
+
+When global is true, list the mixins in the global cache:
+
+```console
+$ porter mixins list [--global=true]
+Name         Versions
+exec         12
+helm         3
+```
+
+The json formatted output should list all versions in the global cache.
+
+
+#### porter mixins install
+
+When global is false, update the porter.yaml file to declare the specified mixin, install the mixin and cache it in the bundle mixin cache.
+This mode is useful for automating the creation of bundles.
+
+When global is true, install the mixin into the global bundle cache (current behavior).
+This mode is useful for pre-populating the global mixin cache.
+
```

Comment:
For a bundle with dependencies, do we need to resolve the dependents mixins? If we have to have a mechanism to resolve them and install them as well as deciding on the mixin version to install.
Also if a dependent mixin is not fulfilled, do we need to break?

Reply:
Dependencies are not resolved or used at build time so I don't think we need to worry about them here? By the time you are adding a reference to a dependency in your porter.yaml, that dependency has already been built and published to a registry.